# TIMEIT

In [1]:
import timeit

# x = 2 + 2
print(timeit.timeit('x = 2 + 2'))
print(timeit.timeit('x = sum(range(10))'))

0.020605999999999902
0.3161035000000001


In [2]:
import cProfile

def get_len(array):
    return len(array)

def get_sum(array):
    s = 0
    for i in array:
        s += i
    return s

def main():
    my_list = [i for i in range(10000000)]
    a = get_len(my_list)
    b = get_sum(my_list)

In [3]:
%%timeit
main()

1.05 s ± 36.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [4]:
cProfile.run('main()')

         8 function calls in 1.064 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.964    0.964 <ipython-input-2-486a2b3e0764>:12(main)
        1    0.520    0.520    0.520    0.520 <ipython-input-2-486a2b3e0764>:13(<listcomp>)
        1    0.000    0.000    0.000    0.000 <ipython-input-2-486a2b3e0764>:3(get_len)
        1    0.444    0.444    0.444    0.444 <ipython-input-2-486a2b3e0764>:6(get_sum)
        1    0.100    0.100    1.064    1.064 <string>:1(<module>)
        1    0.000    0.000    1.064    1.064 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




# РЕКУРСИВНЫЙ ПОИСК ЧИСЕЛ ФИБОНАЧЧИ

In [5]:
import cProfile

def test_fib(func):
    my_list = [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
    for i, item in enumerate(my_list):
        assert item == func(i)
        print(f'test {i} is OK')

def fib(n):
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)

test_fib(fib)

test 0 is OK
test 1 is OK
test 2 is OK
test 3 is OK
test 4 is OK
test 5 is OK
test 6 is OK
test 7 is OK
test 8 is OK
test 9 is OK


In [6]:
%%timeit -n 1000
fib(10)

20.4 µs ± 465 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [7]:
%%timeit -n 1000
fib(20)

2.52 ms ± 467 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [8]:
%%time
fib(30)

Wall time: 298 ms


832040

In [9]:
cProfile.run('fib(10)')

         180 function calls (4 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    177/1    0.000    0.000    0.000    0.000 <ipython-input-5-074ac247022c>:9(fib)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [10]:
cProfile.run('fib(20)')

         21894 function calls (4 primitive calls) in 0.006 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  21891/1    0.006    0.000    0.006    0.006 <ipython-input-5-074ac247022c>:9(fib)
        1    0.000    0.000    0.006    0.006 <string>:1(<module>)
        1    0.000    0.000    0.006    0.006 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




#### как видно, при увеличении числа Фибоначчи в 2 раза (с 10 до 20) количество вызовов функции увеличилось со 180 до 21894 и время с 0 до 0.005 seconds

# РЕКУРСИВНЫЙ ПОИСК ЧИСЕЛ ФИБОНАЧЧИ С МЕМОИЗАЦИЕЙ

### используем словарь

In [11]:
import cProfile


def test_fib(func):
    my_list = [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
    for i, item in enumerate(my_list):
        assert item == func(i)
        print(f'test {i} is OK')
        
        
def fib_dict(n):
    fib_d = {0:0, 1:1}
    
    def _fib_dict(n):
        if n in fib_d:
            return fib_d[n]
        
        fib_d[n] = _fib_dict(n - 1) + _fib_dict(n - 2)
        return fib_d[n]
    
    return _fib_dict(n)

test_fib(fib_dict)

test 0 is OK
test 1 is OK
test 2 is OK
test 3 is OK
test 4 is OK
test 5 is OK
test 6 is OK
test 7 is OK
test 8 is OK
test 9 is OK


In [12]:
%%timeit -n 1000
fib_dict(10)

3.91 µs ± 473 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [13]:
%%timeit -n 1000
fib_dict(20)

7.67 µs ± 429 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [14]:
%%time
fib_dict(30)

Wall time: 0 ns


832040

In [15]:
cProfile.run('fib_dict(10)')

         23 function calls (5 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <ipython-input-11-463280140c08>:11(fib_dict)
     19/1    0.000    0.000    0.000    0.000 <ipython-input-11-463280140c08>:14(_fib_dict)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [16]:
cProfile.run('fib_dict(20)')

         43 function calls (5 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <ipython-input-11-463280140c08>:11(fib_dict)
     39/1    0.000    0.000    0.000    0.000 <ipython-input-11-463280140c08>:14(_fib_dict)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




#### как видно, при увеличении числа Фибоначчи в 2 раза (с 10 до 20) количество вызовов функции увеличилось с 23 до 43 

In [17]:
cProfile.run('fib_dict(100)')

         203 function calls (5 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <ipython-input-11-463280140c08>:11(fib_dict)
    199/1    0.000    0.000    0.000    0.000 <ipython-input-11-463280140c08>:14(_fib_dict)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




#### как видно, при увеличении числа Фибоначчи в 10 раз (с 10 до 100) количество вызовов функции увеличилось с 23 до 203, время изменилось незначительно

### используем список

In [18]:
import cProfile


def test_fib(func):
    my_list = [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
    for i, item in enumerate(my_list):
        assert item == func(i)
        print(f'test {i} is OK')
        
        
def fib_list(n):
    fib_l = [None] * 1000
    fib_l[0:2] = [0, 1]
    
    def _fib_list(n):
        if fib_l[n] is None:
            fib_l[n] = _fib_list(n - 1) + _fib_list(n - 2)
        return fib_l[n]
    
    return _fib_list(n)

test_fib(fib_list)

test 0 is OK
test 1 is OK
test 2 is OK
test 3 is OK
test 4 is OK
test 5 is OK
test 6 is OK
test 7 is OK
test 8 is OK
test 9 is OK


In [19]:
%%timeit -n 1000
fib_list(10)

8.32 µs ± 481 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [20]:
%%timeit -n 1000
fib_list(20)

11.3 µs ± 471 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [21]:
%%time
fib_list(30)

Wall time: 0 ns


832040

In [22]:
cProfile.run('fib_list(10)')

         23 function calls (5 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <ipython-input-18-016bf45b3c08>:11(fib_list)
     19/1    0.000    0.000    0.000    0.000 <ipython-input-18-016bf45b3c08>:15(_fib_list)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [23]:
cProfile.run('fib_list(20)')

         43 function calls (5 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <ipython-input-18-016bf45b3c08>:11(fib_list)
     39/1    0.000    0.000    0.000    0.000 <ipython-input-18-016bf45b3c08>:15(_fib_list)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [24]:
cProfile.run('fib_list(100)')

         203 function calls (5 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <ipython-input-18-016bf45b3c08>:11(fib_list)
    199/1    0.000    0.000    0.000    0.000 <ipython-input-18-016bf45b3c08>:15(_fib_list)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




# НЕРЕКУРСИВНЫЙ ПОИСК ЧИСЕЛ ФИБОНАЧЧИ

In [25]:
import cProfile

def test_fib(func):
    my_list = [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
    for i, item in enumerate(my_list):
        assert item == func(i)
        print(f'test {i} is OK')
        
def fib_loop(n):
    if n < 2:
        return n
    
    first, second = 0, 1
    for i in range(2, n + 1):
        first, second = second, first + second
    
    return second

In [26]:
test_fib(fib_loop)

test 0 is OK
test 1 is OK
test 2 is OK
test 3 is OK
test 4 is OK
test 5 is OK
test 6 is OK
test 7 is OK
test 8 is OK
test 9 is OK


In [27]:
%%timeit -n 1000
fib_loop(10)

855 ns ± 207 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [29]:
%%timeit -n 1000
fib_loop(100)

5.6 µs ± 499 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [32]:
%%timeit -n 1000
fib_loop(2000)

149 µs ± 2.52 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


с классической рекурсией такое сделать не получится, т.к. переполнится стэк (лимит в 1000)!

In [33]:
cProfile.run('fib_loop(10)')

         4 function calls in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <ipython-input-25-0627844883b1>:9(fib_loop)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [34]:
cProfile.run('fib_loop(10)')

         4 function calls in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <ipython-input-25-0627844883b1>:9(fib_loop)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [35]:
cProfile.run('fib_loop(1000)')

         4 function calls in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <ipython-input-25-0627844883b1>:9(fib_loop)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




# ИСПОЛЬЗОВАНИЕ ДЕКОРАТОРА @functools.lru_cache()

In [70]:
import cProfile
import functools


def test_fib(func):
    my_list = [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
    for i, item in enumerate(my_list):
        assert item == func(i)
        print(f'test {i} is OK')

@functools.lru_cache()  # данный декоратор хранит входные и выходные данные в памяти 
                        # и в случае запроса с аналогичными параметрами выдаёт ответ сразу 
                        # из памяти без обстукивания его в функции
def fib(n):
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)

In [59]:
%%timeit -n 1000
fib(10)

120 ns ± 4.67 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [60]:
%%timeit -n 1000
fib(20)

85.2 ns ± 1.31 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [61]:
%%timeit -n 1000
fib(100)

79 ns ± 13 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [62]:
%%timeit -n 1000
fib(5000)

RecursionError: maximum recursion depth exceeded in comparison

## декоратор даёт значительные прирост в скорости, но не спасает от переполнения стэка!!!

In [71]:
cProfile.run('fib(10)')

         14 function calls (4 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     11/1    0.000    0.000    0.000    0.000 <ipython-input-70-88f1a7cf83cb>:11(fib)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [72]:
cProfile.run('fib(20)')

         13 function calls (4 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     10/1    0.000    0.000    0.000    0.000 <ipython-input-70-88f1a7cf83cb>:11(fib)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [73]:
cProfile.run('fib(100)')

         83 function calls (4 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     80/1    0.000    0.000    0.000    0.000 <ipython-input-70-88f1a7cf83cb>:11(fib)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [74]:
cProfile.run('fib(200)')

         103 function calls (4 primitive calls) in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    100/1    0.000    0.000    0.000    0.000 <ipython-input-70-88f1a7cf83cb>:11(fib)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


